src/iocp/mod.rs

```diff
@@ -0,0 +1,821 @@
+            self.drain_update_queue(false)?;
+
+            // Get the time to wait for.
+            let timeout = deadline.map(|t| t.saturating_duration_since(Instant::now()));
+
+            // Wait for I/O events.
+            let len = self.port.wait(&mut packets, timeout)?;
+            log::trace!("new events: handle={:?}, len={}", self.port, len);
+
+            // We are no longer polling.
+            drop(guard);
+
+            // Process all of the events.
+            for entry in packets.drain(..) {
+                let packet = entry.into_packet();
+
+                // Feed the event into the packet.
+                match packet.feed_event(self)? {
+                    FeedEventResult::NoEvent => {}
+                    FeedEventResult::Event(event) => {
+                        events.packets.push(event);
+                        new_events += 1;
+                    }
+                    FeedEventResult::Notified => {
+                        notified = true;
+                    }
+                }
+            }
+
+            // Break if there was a notification or at least one event, or if deadline is reached.
+            let timeout_is_empty =
+                timeout.map_or(false, |t| t.as_secs() == 0 && t.subsec_nanos() == 0);
+            if notified || new_events > 0 || timeout_is_empty {
+                break;
+            }
+
+            log::trace!("wait: no events found, re-entering polling loop");
+        }
+
+        Ok(())
+    }
+
+    /// Notify this poller.
+    pub(super) fn notify(&self) -> io::Result<()> {
+        // Push the notify packet into the IOCP.
+        self.port.post(0, 0, self.notifier.clone())
+    }
+
+    /// Run an update on a packet.
+    fn update_packet(&self, mut packet: Packet) -> io::Result<()> {
+        loop {
+            // If we are currently polling, we need to update the packet immediately.
+            if self.polling.load(Ordering::Acquire) {
+                packet.update()?;
+                return Ok(());
+            }
+
+            // Try to queue the update.
+            match self.pending_updates.push(packet) {
+                Ok(()) => return Ok(()),
+                Err(p) => packet = p.into_inner(),
+            }
+
+            // If we failed to queue the update, we need to drain the queue first.
+            self.drain_update_queue(true)?;
+        }
+    }
+
+    /// Drain the update queue.
+    fn drain_update_queue(&self, limit: bool) -> io::Result<()> {
+        let max = if limit {
+            self.pending_updates.capacity().unwrap()
+        } else {
+            std::usize::MAX
+        };
+
+        // Only drain the queue's capacity, since this could in theory run forever.
+        for _ in 0..max {
+            if let Ok(packet) = self.pending_updates.pop() {
+                packet.update()?;
+            } else {
+                return Ok(());
+            }
+        }
```

Comment:
```suggestion
        core::iter::from_fn(|| self.pending_updates.pop().ok())
            .take(max)
            .map(|packet| packet.update())
            .collect::<Result<Vec<()>,_>>()?;
```
the `Vec<()>` shouldn't actually allocate anything, as `()` is ZST.

Reply:
Great idea! I used `try_for_each` instead of `.map.collect` since it's more idiomatic.

Unrelated: Should we implement an `Iterator` for draining a `ConcurrentQueue` in order to avoid using `from_fn()`?

Reply:
would be ok, it should be documented that the `Iterator::next`, analogous to the rest of the interface, won't block.